### ISSUE TITLE
[C#] Implement new Concept Exercise: object-intializers

### ISSUE THREAD

**Issue by mikedamay** [exercism/v3#1637]
This issue describes how to implement the `object-initializers` concept exercise for the C# track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [The features of v3](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md).
- [Rationale for v3](https://github.com/exercism/v3/blob/master/docs/rationale-for-v3.md).
- [What are concept exercise and how they are structured?](https://github.com/exercism/v3/blob/master/docs/features-of-v3.md)

Please also watch the following video:

- [The Anatomy of a Concept Exercise](https://www.youtube.com/watch?v=gkbBqd7hPrA).

## Goal

The goal of this exercise is to teach the student how and when to use object initializers in C#.

## Learning objectives

- Know how and when to user object initializers.
- Understand that object initializers are used with collections (discussion only).

## Out of scope

- Anonymous classes

## Concepts

- `object-initializers`

## Prerequisites

- `constructors`

Other prerequisites will depend on the story.

### Hints
[object initializer][object-initializers] documentation describes how to use initialzers.

### After

[object initializer][object-initializers] documentation describes how to use initialzers.
[collection initializer][collection-initializers] documentation gives a preview of how initializers can be used with collections like dictionaries and lists.

## Representer

Consider adding a feature to normalize dictionary initialization to the [representer][representer].

## Analyzer

This exercise does not require any specific analyzer logic to be added to the [analyzer][analyzer].

## Implementing

To implement this exercise, please [follow these instructions](https://github.com/exercism/v3/blob/master/languages/csharp/reference/implementing-a-concept-exercise.md).

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue.

[object-initializers]: https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/object-and-collection-initializers#object-initializers
[collection-initializers]: https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/object-and-collection-initializers#collection-initializers
[analyzer]: https://github.com/exercism/csharp-analyzer
[representer]: https://github.com/exercism/csharp-representer

**Comment by mikedamay** [comment]
Some class with a number of built in instances available as static members

**Comment by mikedamay** [comment]
I think we should include collections in the code (I was the person that made them discussion only).  We can make `dictionaries` and `lists` pre-requisites.

I'm thinking either some prebuilt wizard warriors or remote control cars.

**Comment by mikedamay** [comment]
> We can make dictionaries and lists pre-requisites.

I think it is better if this is a prerequisite for `dictionaries` and `lists`.  It already is for dictionaries.

**Comment by ErikSchierboom** [comment]
> I think we should include collections in the code (I was the person that made them discussion only). We can make dictionaries and lists pre-requisites.

Agreed! This definitely makes the most sense to me.

> I'm thinking either some prebuilt wizard warriors or remote control cars.

I like it!

**Comment by mikedamay** [comment]
@ErikSchierboom 

> I like it!

Sadly I don't think that w&w/d&d is a fruitful area for me.  And I certainly can't hold my end up in a conversation about cars (remote, racing or otherwise).  User authentication is more my thing.

I was thinking of some hardcoded identities, developers for example.  After tasks covering building the initialization of a non-collection object,  tasks progressively add in a reference member then a list and we finish by having an initialized dictionary of these identities.

I am a little concerned about the number of representations that this may generate.

Any thoughts?

**Comment by mikedamay** [comment]
> another source of inspiration.

I've looked through the Elixir stories in detail and got some inspiration for a couple of other concepts.

In general, I don't find the stories particularly helpful unless they are addressing the same sort of concept as our exercise..

It's the concept-to-language map that would be more help but I appreciate there are other priorities and hitting Elixir has probably covered 80% of the available resources.

I also think stories could be retro-fitted with a few identifier name changes, etc.

I will do an example for the current idea.

**Comment by ErikSchierboom** [comment]
> It's the concept-to-language map that would be more help

I now have the information needed to do this. I will implement something like this later.

> I will do an example for the current idea.

Great.

**Comment by mikedamay** [comment]
@ErikSchierboom  An example of this is in these [instructions](https://github.com/mikedamay/v3/blob/object-initialization/languages/csharp/exercises/concept/object-initialization/.docs/instructions.md) and this [example](https://github.com/mikedamay/v3/blob/object-initialization/languages/csharp/exercises/concept/object-initialization/.meta/Example.cs)

We can add intermediate tasks perhaps:

- Task -1 - store and return admin with non-reference members only
- Task 0 - store and return admin with non-ref + list
and then the current tasks 1 and 2.

**Comment by ErikSchierboom** [comment]
I like that. An alternative could be to have a Task 0 that only initializes the FacialFeatures.

**Comment by ErikSchierboom** [comment]
Oh and the story works!

**Comment by mikedamay** [comment]
closed by PR #1775
